proxy/proxy/client_test.go

```diff
@@ -183,3 +192,54 @@ func TestMaximumConnectionsCount(t *testing.T) {
 		t.Errorf("client should have dialed exactly the maximum of %d connections (%d connections, %d dials)", maxConnections, numConnections, dials)
 	}
 }
+
+func TestRefreshTimer(t *testing.T) {
+	refreshCertBuffer = time.Millisecond * 10
+	timeToExpire := time.Millisecond * 500
+	b := &fakeCerts{}
+	c := &Client{
+		Certs: &blockingCertSource{
+			map[string]*fakeCerts{
+				instance: b,
+			},
+			time.Now().Add(timeToExpire),
+		},
+		Dialer: func(string, string) (net.Conn, error) {
+			return nil, errFakeDial
+		},
+		RefreshCfgThrottle: 20 * time.Millisecond,
+	}
+
+	// Call Dial to cache the cert.
+	if _, err := c.Dial(instance); err != errFakeDial {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	c.cacheL.Lock()
+	cached, ok := c.cfgCache[instance]
+	c.cacheL.Unlock()
+	if !ok {
+		t.Error("expected instance to be cached")
+	}
+	waitTil := time.After(timeToExpire + (10 * time.Millisecond))
+loop:
```

Comment:
Maybe this came from a previous review comment of mine, but: this loop doesn't do anything different than simply executing this sleep:
   
    time.Sleep(timeToExpire + 10*time.Millisecond)

I somewhat recall asking for a sleep+retry loop. If that's what you were attempting to implement here, you should be checking the cfgCache multiple times and sleep+retrying if the config you were looking for was not found/refreshed. That will reduce flakiness by giving the background goroutines some time to process while still reducing the amount of time the test takes in the common case.